### PR TITLE
fix: redirect on register

### PR DESCRIPTION
**Pull request by owlas**
Closes: #7571 

Things to tidy up:
- Refactor redirect code in Login, Register, Invite components - so that all methods use the same redirectURL
- Session data is cleared on `.authenticate` in passport: https://medium.com/passportjs/fixing-session-fixation-b2b68619c51d - are we accounting for this?
- Apply redirect logic consistently across all OIDC providers

Cases to test (common across password and all SSO providers):
- After logging in, I'm redirected to the homepage or `from` if it exists
- After registering via `/register` I'm redirected to the homepage or `from` if it exists
- After registering via `/login` with SSO I'm redirected to the homepage or `from` if it exists
- After registering via an invite link I'm redirected to the hompage or `from` if it exists

### CODE REVIEW

File: packages/frontend/src/pages/Register.tsx [hunk 1]
```diff
@@ -69,6 +70,7 @@ const Register: FC = () => {
                     key={providerName}
                     providerName={providerName}
                     intent="signup"
+                    redirect={redirectUrl}
```
Comment (owlas) [inline packages/frontend/src/pages/Register.tsx:73]: Apply redirect also to SSO login

File: packages/frontend/src/pages/Register.tsx [hunk 2]
```diff
@@ -33,6 +33,9 @@ const Register: FC = () => {
     const allowPasswordAuthentication =
         !health.data?.auth.disablePasswordAuthentication;
     const { identify } = useTracking();
+    const redirectUrl = location.state?.from
```
Comment (owlas) [inline packages/frontend/src/pages/Register.tsx:36]: pull out redirect logic across SSO and password

File: packages/frontend/src/components/common/ThirdPartySignInButton/index.tsx [hunk 3]
```diff
@@ -71,21 +72,17 @@ export const ThirdPartySignInButton: FC<ThirdPartySignInButtonProps> = ({
...
 }) => {
     const { health } = useApp();
-    const location = useLocation<{ from?: Location } | undefined>();
 
     switch (providerName) {
         case OpenIdIdentityIssuerType.GOOGLE:
             return health.data?.auth.google.enabled ? (
                 <ThirdPartySignInButtonBase
                     loginPath={health.data.auth.google.loginPath}
-                    redirect={
```
Comment (owlas) [inline packages/frontend/src/components/common/ThirdPartySignInButton/index.tsx:84]: remove google-specific default

File: packages/frontend/src/components/common/ThirdPartySignInButton/index.tsx [hunk 4]
```diff
@@ -71,21 +72,17 @@ export const ThirdPartySignInButton: FC<ThirdPartySignInButtonProps> = ({
     inviteCode,
     intent = 'signin',
     providerName,
+    redirect,
```
Comment (owlas) [inline packages/frontend/src/components/common/ThirdPartySignInButton/index.tsx:75]: redirect applies to all providers

File: packages/backend/src/controllers/authentication.ts [hunk 5]
```diff
@@ -149,15 +149,15 @@ export const redirectOIDCSuccess: RequestHandler = (req, res) => {
             ? decodeURIComponent(req.query.returnTo.toString())
             : undefined;
 
-    if (queryReturn && queryReturn.startsWith(lightdashConfig.siteUrl)) {
+    if (
+        queryReturn &&
+        new URL(queryReturn, lightdashConfig.siteUrl).host ===
```
Comment (owlas) [inline packages/backend/src/controllers/authentication.ts:154]: allows `queryReturn` to be path only e.g. `/` or `/path/to/chart` without the host

File: packages/backend/src/controllers/authentication.ts [hunk 6]
```diff
@@ -149,15 +149,15 @@ export const redirectOIDCSuccess: RequestHandler = (req, res) => {
...
+            new URL(lightdashConfig.siteUrl).host
+    ) {
         res.redirect(queryReturn);
-        return;
     }
-
-    const { returnTo } = req.session.oauth || {};
```
Comment (owlas) [inline packages/backend/src/controllers/authentication.ts:157]: don't use session redirect anymore (this is redundant now)

File: packages/backend/src/controllers/authentication.ts [hunk 7]
```diff
@@ -129,7 +129,7 @@ export const storeOIDCRedirect: RequestHandler = (req, res, next) => {
     }
     if (typeof redirect === 'string') {
         try {
-            const redirectUrl = new URL(redirect);
+            const redirectUrl = new URL(redirect, lightdashConfig.siteUrl);
```
Comment (owlas) [inline packages/backend/src/controllers/authentication.ts:132]: allows queryReturn to be path only e.g. / or /path/to/chart without the host

**Review by rephus — COMMENTED** [review]
Sorry, this was probably introduced by https://github.com/lightdash/lightdash/pull/7495 ( Not sure what is wrong, I tested this locally :/ ) Feel free to revert my code if you prefer

**Review by rephus — COMMENTED** [review]
You have done a great job applying the change to multiple SSO , and extracting the logic into a separate method. I'm not sure I understand what's the real fix here , is the host prefix ? Were you able to reproduce the redirect issue ?

**Comment by rephus** [comment]
I tested locally and it works fine

**Review by rephus — APPROVED** [review]

**Pull request merged**